reps/2022-04-21-state-observability-apis.md

```diff
@@ -0,0 +1,138 @@
+
+## Summary
+### General Motivation
+
+Ray does not currently expose public APIs for understanding the state of the system (what is running, resource consumption, errors, etc.). This is a frequent pain point for end-users and developers. This REP proposes to introduce structured APIs as part of the Ray 2.0 API for key states: actors, tasks, scheduling info, objects, resources, logs, etc.
+
+#### Key requirements:
+- Need to expose all necessary Ray states for basic observability.
+- State APIs should be consistent / stable / well-documented. E.g., APIs names must be consistent to the concepts
+- APIs should be available to all clients, such as Dashboard, CLI, and Python APIs with the same implementation.
+- Shouldn’t introduce high overhead to the normal workloads.
+- Under no circumstances should it be possible to crash a cluster (say, up to our scalability envelope) using the observability APIs.
+
+
+### Should this change be within `ray` or outside?
+
+Accessing state is a fundamental feature to debug systems. These changes would lie within Ray to improve the basic observability of the project.
+
+## Stewardship
+### Required Reviewers
+
+@scv119, @edoakes, @ericl
+
+### Shepherd of the Proposal (should be a senior committer)
+
+@edoakes
+
+## Design and Architecture
+
+### States
+“States” must be closely related and consistent to “concepts / APIs” users learn. For example, if we expose tasks as core Ray concepts, users would like to see the state of tasks. Ideally, the name of the APIs should be consistent with the concepts too.
+
+State APIs can be categorized by 3 types.
```

Comment:
Are there different requirements for the three categories? Why is it necessary to group them?

Reply:
fine as is